<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial transcriptomics data analysis – spatially-aware clustering with clustSIGNAL • clustSIGNALworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Spatial transcriptomics data analysis – spatially-aware clustering with clustSIGNAL">
<meta property="og:description" content="This workshop serves as a tutorial for using clustSIGNAL (clustering of Spatially Informed Gene expression with Neighbourhood Adapted Learning) - a method for adaptively smoothing and clustering gene expression data obtained from molecule-resolved spatial transcriptomics technologies.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">clustSIGNALworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/clustSIGNAL_tutorial.html">clustSIGNAL tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/SydneyBioX/clustSIGNAL_workshop" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level2">
<h2 id="workshop-overview">Workshop overview<a class="anchor" aria-label="anchor" href="#workshop-overview"></a>
</h2>
<p>In this workshop, we will learn how clustSIGNAL, a spatial cell-type clustering method, works and how to run it on R. There will also be discussion on the various parameters available to us to run the method, how we can assess the relevance of the clusters, and the additional information generated as a by-product that could be used for other analyses.</p>
<p>The clustSIGNAL R package is available to download and install from <a href="https://github.com/SydneyBioX/clustSIGNAL" class="external-link">github/SydneyBioX/clustSIGNAL</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"SydneyBioX/clustSIGNAL"</span><span class="op">)</span></span></code></pre></div>
<p>If you bring your own data, you can run the method locally. See below a list of R packages that clustSIGNAL depends on: <br> SpatialExperiment, BiocNeighbors, BiocParallel, bluster, scater, harmony, rlist</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h2>
<p>The attendees are expected to have:</p>
<ol style="list-style-type: decimal">
<li>Experience in R programming, and</li>
<li>Familiarity with SingleCellExperiment and/or SpatialExperiment objects.</li>
</ol>
<p><br></p>
</div>
<div class="section level2">
<h2 id="content-and-objective">Content and objective<a class="anchor" aria-label="anchor" href="#content-and-objective"></a>
</h2>
<p>Below is a list of tasks we will go through during the workshop, and what the attendees can expect to learn from them.</p>
<table class="table">
<colgroup>
<col width="43%">
<col width="10%">
<col width="46%">
</colgroup>
<thead><tr>
<th>Task</th>
<th>Time</th>
<th>Objective</th>
</tr></thead>
<tbody>
<tr>
<td>Concept behind clustSIGNAL</td>
<td>15 min</td>
<td>Description of how clustSIGNAL works as a spatial clustering method. How does clustSIGNAL</td>
</tr>
<tr>
<td></td>
<td></td>
<td>embed spatial information into the data and how that affects clustering.</td>
</tr>
<tr>
<td>Method parameters</td>
<td>5 min</td>
<td>Learn about the different clustering parameters available in clustSIGNAL, and understand</td>
</tr>
<tr>
<td></td>
<td></td>
<td>how they can affect the clustering output.</td>
</tr>
<tr>
<td>How to run clustSIGNAL</td>
<td>10 min</td>
<td>Perform clustering on spatial transcriptomics data with clustSIGNAL. Also, learn what sort</td>
</tr>
<tr>
<td></td>
<td></td>
<td>of outputs can be obtained from running the method.</td>
</tr>
<tr>
<td>Assessing relevance of clusters</td>
<td>10 min</td>
<td>Investigate the biological relevance of the clusters obtained from running clustSIGNAL.</td>
</tr>
<tr>
<td>Exploring clustSIGNAL outputs</td>
<td>10 min</td>
<td>Learn about how some of the data generated by clustSIGNAL can be useful for other analyses.</td>
</tr>
</tbody>
</table>
<p><br></p>
</div>
<div class="section level2">
<h2 id="id_1-concept-behind-clustsignal">1. Concept behind clustSIGNAL<a class="anchor" aria-label="anchor" href="#id_1-concept-behind-clustsignal"></a>
</h2>
<p>clustSIGNAL (<strong><em>clust</em></strong>ering of <strong><em>S</em></strong>patially <strong><em>I</em></strong>nformed <strong><em>G</em></strong>ene expression with <strong><em>N</em></strong>eighbourhood <strong><em>A</em></strong>dapted <strong><em>L</em></strong>earning) is a cell-type clustering method for subcellular spatial transcriptomics data. It aims to address data sparsity by performing an adaptive smoothing approach to generate modified gene expressions that are embedded with spatial context and neighbourhood composition information.</p>
<p>To capture neighbourhood composition of each cell, we used entropy as a measure of the “domainness” of cell neighbourhoods - the more homogeneous a neighbourhood the lower its entropy and the more heterogeneous a neighbourhood the higher its entropy. The entropy values were then used to generate cell neighbourhood-specific weights to perform an adaptive smoothing of gene expression, such that smoothing was performed over more cells in homogeneous neighbourhoods, but heterogeneous neighbourhoods were smoothed over a much smaller region.</p>
<p><br></p>
<p><img src="inst/images/Fig1_schematic.jpg" align="middle"></p>
<p><br></p>
<p><strong>Figure: clustSIGNAL method overview.</strong></p>
<p>The core steps involved in the method are sequential:</p>
<p><strong>1.</strong> The method starts with non-spatial clustering and subclustering (default louvain clustering) to classify cells into subclusters that we refer to as “intial clusters”. <br><em><strong>Function:</strong> clustSIGNAL::p1_clustering()</em></p>
<p><strong>2.</strong> The neighbourhood of each cell is defined in terms of their “intial clusters” composition. <br><em><strong>Function:</strong> clustSIGNAL::neighbourDetect()</em></p>
<p><strong>3.</strong> The cells in the neighbourhood are also sorted and rearranged so that the neighbours belonging to the same “intial clusters” group as the index cell are placed closer to it. <br><em><strong>Function:</strong> clustSIGNAL::neighbourDetect()</em></p>
<p><strong>4.</strong> Neighbourhood “domainness” is measured as entropy, where high entropy values indicate more heterogeneous neighbourhoods and low entropy values indicate more homogeneous neighbourhoods. <br><em><strong>Function:</strong> clustSIGNAL::entropyMeasure()</em></p>
<p><strong>5.</strong> The entropy values are used to generate weight distributions specific to each neighbourhood. <br><em><strong>Function:</strong> clustSIGNAL::adaptiveSmoothing()</em></p>
<p><strong>6.</strong> The gene expressions of cells are adaptively smoothed using the entropy-guided weight distributions; cells in heterogeneous neighbourhoods (high entropy) undergo smoothing over a smaller region, whereas cells in homogeneous neighbourhoods (low entropy) undergo smoothing over a larger region. <br><em><strong>Function:</strong> clustSIGNAL::adaptiveSmoothing()</em></p>
<p><strong>7.</strong> Non-spatial clustering is performed with adaptively smoothed gene expression to generate clustSIGNAL clusters that represent cell types. <br><em><strong>Function:</strong> clustSIGNAL::p2_clustering()</em></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="id_2-method-parameters">2. Method parameters<a class="anchor" aria-label="anchor" href="#id_2-method-parameters"></a>
</h2>
<p>The clustSIGNAL package uses a SpatialExperiment object as input. We provide users with a number of parameters to explore and experiment with, as well as prior tested default values for quick runs. clustSIGNAL can be used for single sample or multisample analysis with just one function call. Below is the list of the parameters offered and their possible values:</p>
<ul>
<li><p><strong>spe</strong> - SpatialExperiment object with cell <a href="#id_#" title="Stored in the spatialCoords(spe) cell location section">spatial coordinates</a> matrix and <a href="#id_#" title="Stored under logcounts(spe) assay">normalized counts</a> of gene expression.</p></li>
<li><p><strong>samples</strong> - column name in <a href="#id_#" title="Stored in the colData(spe) section">cell metadata</a> containing sample names.</p></li>
<li><p><strong>cells</strong> - column name in <a href="#id_#" title="Stored in the colData(spe) section">cell metadata</a> containing cell IDs.</p></li>
<li><p><strong>dimRed</strong> - dimensionality reduction method name in <a href="#id_#" title="Stored in the reducedDimNames(spe) embeddings section">low embedding data</a>. Default value is “None”, in which case PCA is calculated and used as low dimension data.</p></li>
<li><p><strong>batch</strong> - whether batch correction should be performed. Default value is FALSE.</p></li>
<li><p><strong>batch_by</strong> - column name in <a href="#id_#" title="Stored in the colData(spe) section">cell metadata</a> containing the groups by which to perform batch correction.</p></li>
<li><p><strong>NN</strong> - neighbourhood size in terms of the number of nearest neighbours to consider. Value should be &gt; 1. Default value is 30.</p></li>
<li><p><strong>kernel</strong> - type of weight distribution to use. Can be Gaussian (default) or exponential distribution.</p></li>
<li><p><strong>spread</strong> - value of distribution parameter - standard deviation of Gaussian distribution or rate of exponential distribution. Default value is 0.05, recommended for Gaussian distribution. For exponential distribution, recommended value is 20.</p></li>
<li><p><strong>sort</strong> - whether cell neighbourhoods should be sorted by their ‘putative cell type’ grouping. Default value is True.</p></li>
<li><p><strong>threads</strong> - number of cpus to use for parallel runs. Default value is 1.</p></li>
<li><p><strong>outputs</strong> - choice of output types. Default value is ‘c’ for data frame of cell IDs and cluster numbers. Other possible value is “a” for a list of dataframe of clusters plus final SpatialExperiment object.</p></li>
<li><p><strong>clustParams</strong> - parameter options for TwoStepParam clustering methods in the bluster package. The clustering parameters are in the order - centers (centers) for clustering with KmeansParam, centers (centers) for sub-clustering clusters with KmeansParam, maximum iterations (iter.max) for clustering with KmeansParam, k values (k) for clustering with NNGraphParam, and community detection method (cluster.fun) to use with NNGraphParam.</p></li>
</ul>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/SydneyBioX/clustSIGNAL_workshop/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/SydneyBioX/clustSIGNAL_workshop/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing clustSIGNALworkshop</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Pratibha Panwar <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-7437-7084" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Shila Ghazanfar <br><small class="roles"> Author, contributor </small> <a href="https://orcid.org/0000-0001-7861-6997" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Pratibha Panwar, Shila Ghazanfar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
