## Overview

clustSIGNAL is an R package that performs spatially-resolved clustering on spatial transcriptomics data. It uses an entropy-guided approach to measure the "domainness" of cell neighbourhoods - homogeneous neighbourhoods have low entropy and heterogeneous neighbourhoods have high entropy. The entropy values are used to generate cell neighbourhood-specific weights to perform an adaptive smoothing of gene expression - smoothing is performed over more cells in homogeneous neighbourhoods, but heterogeneous neighbourhoods are smoothed over a much smaller region. This approach aims to overcome data sparsity in spatial transcriptomics gene expression, while incorporating spatial context in the form of cell arrangement information of each cell's neighbourhood.



## Pre-requisites to run the workshop

The attendees are expected to have:

1. Experience with R programming, and
2. Familiarity with SingleCellExperiment and/or SpatialExperiment objects.



## Content and objective
Below is a list of tasks we will go through during the workshop, and what the attendees can expect to learn from them.

| Task                                  | Time   | Objective   |
|---------------------------------------|--------|-------------|
| Concept behind clustSIGNAL            | 15 min | Description of how clustSIGNAL works as a spatial clustering method. How does clustSIGNAL embed spatial information into the data and how that affects clustering.
| Method parameters                     | 5 min  | Learn about the different clustering parameters available in clustSIGNAL, and understand how they can affect the clustering output. 
| How to run clustSIGNAL                | 10 min | Perform clustering on spatial transcriptomics data with clustSIGNAL. Also, learn what sort of outputs can be obtained from running the method.
| Assessing relevance of clusters       | 10 min | Investigate the biological relevance of the clusters obtained from running clustSIGNAL.
| Exploring clustSIGNAL outputs         | 10 min | Learn about how some of the data generated by clustSIGNAL can be useful for other analyses.



## Notes

The clustSIGNAL R package is available to download and install from [github/SydneyBioX/clustSIGNAL](https://github.com/SydneyBioX/clustSIGNAL).

The attendees can bring their own data, and run the method locally. 
