---
title: "clustSIGNAL tutorial"
author:
  - Pratibha Panwar, Boyi Guo, Haowen Zhou, Stephanie Hicks, Shila Ghazanfar
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{clustSIGNAL tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
editor_options: 
  markdown: 
    wrap: 80
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  comment = "#>",
  cache = FALSE,
  message = FALSE,
  warning = FALSE
)
```

<style>
.question {
  padding: 1em;
  background: lightcyan;
  color: black;
  border-radius: 10px;
}
</style>

**Presenter**

Pratibha Panwar$^{1,2,3}$

$^1$ School of Mathematics and Statistics, University of Sydney, Australia\
$^2$ Sydney Precision Data Science Centre, University of Sydney, Australia\
$^3$ Charles Perkins Centre, University of Sydney, Australia

Contact: pratibha.panwar\@sydney.edu.au

In this RStudio instance, we will cover the remaining tasks.

### Setup Installation

```{r}
# you do not need to run this chunk
# the packages are already installed in this instance

# install from CRAN
# install.packages(c("Seurat", "aricode", "dplyr", "ggplot2", "patchwork"))

# install from github
# install.packages("devtools")
# devtools::install_github("SydneyBioX/clustSIGNAL")
```

### Loading packages

```{r load_packages, message = FALSE, warning = FALSE}
# load required packages
library(clustSIGNAL)
library(Seurat)
library(aricode)
library(dplyr)
library(ggplot2)
library(patchwork)
```

# 3. How to run clustSIGNAL

## Example datasets

For this tutorial, we will use sampled data from two publicly available
datasets - (i) a subsample from the SeqFISH mouse embryo dataset from [Lohoff et
al, 2021](https://www.nature.com/articles/s41587-021-01006-2), and (ii) a
subsample from the MERFISH mouse hypothalamus preoptic region dataset from
[Moffitt et al, 2018](https://www.science.org/doi/10.1126/science.aau5324)

The SeqFISH mouse embryo dataset contains gene expression data for 3 mouse
embryos, with 351 genes and a total of 57,536 cells. Here, we use a subset data
of 5000 randomly selected cells from Embryo 2, excluding cells that had been
annotated as 'Low quality'.

The MERFISH mouse hypothalamus preoptic region dataset contains gene expression
for 181 samples, with 155 genes and a total of 1,027,080 cells. Here, we use a
subset data of 6000 cells randomly selected from only 3 samples - Animal 1
Bregma -0.09 (2080 cells), Animal 7 Bregma 0.16 (1936 cells), and Animal 7
Bregma -0.09 (1984 cells), excluding cells that had been annotated as
'Ambiguous' and 20 genes that were assessed using a different technology.

These sampled datasets are available with clustSIGNAL package and can be
accessed as below:

```{r embryo_data}
data(mEmbryo2) 
# this will load me_expr and me_data objects, containing gene expression 
# logcounts and cell metadata, respectively, to your environment
```

```{r hypothal_data}
data(mHypothal) 
# this will load mh_expr and mh_data objects, containing gene expression 
# logcounts and cell metadata, respectively, to your environment
```

## Creating SpatialExperiment objects

ClustSIGNAL requires a SpatialExperiment (spe) object as input, so first we need
to create a spe object from the gene expression and cell metadata we have in our
environment.

The gene expression data is in a gene by cell matrix. The cell metadata is a
dataframe of cell characteristics, including the x-y coordinates of each cell.

Let us work with the MERFISH data first, by creating a spe object from the data.

```{r hypothal_data_prep}
# to create spe object we need gene expression, cell metadata, and cell locations
spe_mh <- SpatialExperiment(assays = list(logcounts = mh_expr),
                            colData = mh_data, 
                            # spatialCoordsNames requires column names in 
                            # mh_data that contain xy-coordinates of cells
                            spatialCoordsNames = c("X", "Y"))
spe_mh
```

Some components of the spe object that are frequently used include assays (the
gene expression data are stored here - multiple assays can be stored in one spe
object), colData (the cell characteristics are stored here), rowData (the gene
characteristics can be stored here), spatialCoords (the x-y coordinates of each
cell are stored here), reducedDims (any low embeddings data can be stored here),
and imgData (any images from the dataset can be stored here).

## Running clustSIGNAL

ClustSIGNAL comes with many parameters that can be explored. Most of these
parameters have default values and do not need to be specified when running the
method, except three parameters that need input. These include the spe object,
the colData column name containing sample names, and the colData column name
containing cell IDs. Moreover, the spe object must have logcounts stored in
assays section and cell locations stored in the spatialCoords section.

Since we created the spe object earlier, we know that the logcounts and spatial
coordinates are stored in the spe object. We can confirm this by just looking at
the spe object.

```{r hypothal_dataCheck}
spe_mh
```

Now, let us identify the column names containing cell IDs and sample names so we
can use them for running clustSIGNAL.

```{r hypothal_cell_column}
spe_mh |> colData() |> head()
```

Here, the cell IDs are in the column named "Cell_ID" and the samples are in the
"samples" column.

Now that we have the column names, and the spe object contains the relevant
information, we can run clustSIGNAL with some default values. Here, clustSIGNAL
will be run on the input spe object, with cell ID and sample names specified by
the column names in samples and cells respectively. Keeping outputs = "a" will
get us the final spe object as well as the dataframe of cluster numbers.

Feel free to explore the data by modifying the input parameters. The **dimRed**
parameter allows the use of specific low embeddings in the spe object for
generation of the initial clusters. The **batch** and **batch_by** parameters
can be used to specify whether to perform batch correction and what sample
groups to use. By default, clustSIGNAL uses information from 30 nearest
neighbours, but this can be modified by changing the value of the **NN**
parameter. To change the weights used for adaptive smoothing, the **kernel** and
**spread** parameters that control the type and spread of distribution,
respectively, can be modified. Clustering parameters can be explored using the
**clustParams** parameter list.

```{r clustSIGNAL_run}
set.seed(110)
smp_label <- "samples" # column name containing sample names
c_label <- "Cell_ID" # column name containing cell IDs
res_mh <- clustSIGNAL(spe_mh, samples = smp_label, cells = c_label, 
                      outputs = "a")
```

Since, we selected all output option, clustSIGNAL returns a list including a
dataframe of cluster numbers and a spe object.

```{r clustSIGNAL_output}
res_mh |> names()
```

The resulting spe object contains the adaptively smoothed gene expression data
as an additional assay, initial clusters, entropy values, and clustSIGNAL
clusters. Essentially, the final spe object contains data from the input spe
object plus the outputs from the clustSIGNAL run.

```{r hypothal_speFinal}
res_mh$spe_final
spe_mh <- res_mh$spe_final
```

# 4. Assessing relevance of clusters

In this section, we will analyse the results from clustSIGNAL through clustering
metrics and visualisations.

Clustering metrics such as ARI (adjusted rand index) and NMI (normalised mutual
information) allow us to compare clustering performed by two methods. Here, we
use ARI and NMI to compare clustSIGNAL clusters with published annotations -
higher values mean more consensus between the two groups.

```{r hypothal_clusterMetrics}
samplesList <- spe_mh[[smp_label]] |> unique() # get unique sample names

spe_mh |> colData() %>% 
  as.data.frame() %>%
  group_by(get(smp_label)) %>% 
  # group cells by samples and for cells in each sample 
  # calculate the following metrics
  summarise(ARI = aricode::ARI(Cell_class, clustSIGNAL), # calculate ARI
            NMI = aricode::NMI(Cell_class, clustSIGNAL)) # calculate NMI
```

The clustering output can also be visualised by plotting the spatial coordinates
of cells and colouring by cluster numbers. Here, we use the plotReducedDim
function from the scater R package for creating spatial plots. This requires the
spatial coordinates to be added to the reducedDimension section of the spe
object.

```{r hypothal_addCoords}
# add the spatial coordinates to the reduced dimension section 
# for plotting with scater package
reducedDim(spe_mh, "spatial") <- spatialCoords(spe_mh)
```

To specify that we want to create spatial plots, we just mention the correct low
embedding name in the dimred option, in this case it is "spatial".

```{r hypothal_spatialPlots}
scater::plotReducedDim(spe_mh, 
                       # specify spatial low dimension
                       dimred = "spatial", 
                       colour_by = "clustSIGNAL", 
                       point_alpha = 1, 
                       point_size = 1) +
  # to separate out the 3 samples in the dataset
  facet_wrap(vars(spe_mh[[smp_label]]), scales = "free")
```

Here, the x and y axes are the x-y coordinates of the cells. The dataset
contains 3 samples, so we have one plot per sample. ClustSIGNAL performs
clustering using all cells in the dataset in one run, thereby identifying the
same clusters across multiple samples in the dataset.

We can also compare the clustering output with the annotated labels to assess
how much of the manual annotation was captured by the method output.

```{r hypothal_clusterCompare}
table(spe_mh$Cell_class, spe_mh$clustSIGNAL)
```

In this table, the rows show published manual annotations and columns show
clustSIGNAL cluster labels. ClustSIGNAL is able to capture the distinct cell
types and also identify subgroups in some cases, e.g., the inhibitory neurons.

To assess how distinct these clusters are, we investigate the top marker genes
in each cluster using the FindAllMarkers function in Seurat R package.

```{r hypothal_seuObject}
# we can convert the spe object into a Seurat (seu) object as below
# since we do not have counts data, we've used logcounts as 
# counts to be able to create the seu object
seu_mh <- Seurat::as.Seurat(spe_mh, data = NULL, counts = "logcounts")
seu_mh
```

Since we've used the logcounts as counts in this seu object, the two assay
layers - counts and data - contain the same information. For finding cluster
markers, we will use default values, which includes using the data layer
containing logcounts.

```{r hypothal_clusterMarkers}
# to specify that the cluster labels are in clustSIGNAL column
Idents(seu_mh) <- "clustSIGNAL"
# this will identify marker genes in each cluster using default values
markers_mh <- Seurat::FindAllMarkers(seu_mh)
```

We can now plot the top marker genes and their gene expression in each cell to
see how distinct each cluster is.

```{r hypothal_expHeatmap}
# here we select the top 10 markers genes (with log fold change > 0.5) 
# from each cluster, but you can modify the parameters to select 
# more/less genes with higher/lower fold change
top10 <- markers_mh %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 0.5) %>%
  slice_head(n = 10) %>%
  ungroup()

# plotting the heatmap for top 10 marker genes using their logcounts ("data")
Seurat::DoHeatmap(seu_mh, slot = "data", features = top10$gene) + NoLegend()
```

Here, the genes are shown along the y-axis and the cells, grouped by the cluster
they belong to, are displayed along the x-axis. The values in the heatmap are
logcounts of top 10 marker genes in each cluster. The heatmap shows that the 
clusters associated with inhibitory neurons have different gene expression 
patterns, which accounts for their separation by clustSIGNAL.

# 5. Exploring clustSIGNAL outputs

Other than the cluster labels, clustSIGNAL also generates smoothed gene
expression, initial cluster labels, and cell neihbourhood-specific entropy
values. Other outputs such as nearest neighbour matrix and initial cluster
based neighbourhood compositions are also accessible by running clustSIGNAL
functions sequentially.

The *p1_clustering()* function generates the initial cluster labels and adds them
to the spe object, the *neighbourDetect()* function produces nearest neighbour
matrix and neighbourhood compositions, the *entropyMeasure()* function adds
entropy values to the spe object, the *adaptiveSmoothing()* function adds smoothed
gene expression to the spe object, and the *p2_clustering()* function generates
the final clusters and adds them to the spe object.

Of these additional outputs, the entropy values can be valuable in exploring
datasets to get a sense of the neighbourhood compositions of the cells,
essentially revealing gene expression based tissue structure. Let us explore
this with the two datasets available to us. First, we need to run clustSIGNAL on
the seqFISH dataset to generate the appropriate data.

```{r embryo_dataPrep_run}
# to create spe object we need gene expression, cell metadata, and cell location
spe_me <- SpatialExperiment(assays = list(logcounts = me_expr),
                            colData = me_data, 
                            # spatialCoordsNames requires column names in 
                            # me_data that contain xy-coordinates of cells
                            spatialCoordsNames = c("X", "Y"))
reducedDim(spe_me, "spatial") <- spatialCoords(spe_me)

set.seed(110)
# "sample_id" column contains sample names
# "uniqueID" column contains cell IDs
res_me <- clustSIGNAL(spe_me, samples = "sample_id", cells = "uniqueID", 
                      outputs = "a")
spe_me <- res_me$spe_final
spe_me
```

To explore the dataset, we can estimate the spread and distribution of the
entropy values using histogram and spatial plots, respectively.

```{r compare_hist}
# histogram plots to show entropy spread
h1 <- spe_me |> colData() |> as.data.frame() %>% # plotting embryo data
  ggplot(aes(entropy)) +
  geom_histogram(binwidth = 0.05) +
  labs(x = "Entropy", y = "Number of regions") +
  theme_classic() +
  theme(text = element_text(size = 12))

h2 <- spe_mh |> colData() |> as.data.frame() %>% # plotting hypothalamus data
  ggplot(aes(entropy)) +
  geom_histogram(binwidth = 0.05) +
  # separate out the 3 samples in the dataset
  facet_wrap(vars(get(smp_label)), nrow = 1) +
  labs(x = "Entropy", y = "Number of regions") +
  theme_classic() +
  theme(text = element_text(size = 12))

# spatial plots to show entropy distribution
s1 <- scater::plotReducedDim(spe_me, # plotting embryo data
                             # specify spatial low dimension
                             dimred = "spatial", 
                             # specify colour points by entropy
                             colour_by = "entropy", 
                             point_alpha = 1, 
                             point_size = 1)
s2 <- scater::plotReducedDim(spe_mh, # plotting hypothalamus data
                             # specify spatial low dimension
                             dimred = "spatial", 
                             # specify colour points by entropy
                             colour_by = "entropy", 
                             point_alpha = 1, 
                             point_size = 1) +
  # to separate out the 3 samples in the dataset
  facet_wrap(vars(spe_mh[[smp_label]]), scales = "free") 

(h1 + h2 + patchwork::plot_layout(widths = c(1, 3))) /  
  (s1 + s2 + patchwork::plot_layout(widths = c(1, 3)))
```

The entropy plots help us gauge the "domainness" in the samples. For example,
the neighbourhood entropy values in the mouse embryo data can go as low as 0,
which would mean that this sample has many neighbouhoods that are completely
homogeneous and could represent domains composed of one celltype. The spatial
plot helps visualise where in the sample such homogeneous regions might be
found. On the contrary, the neighbourhood entropy values in the mouse
hypothalamus data are generally quite high, with no regions that are completely
homogeneous.

One thing to note is that these entropy values are based on 30 nearest
neighbours in this example, and changing the **NN** parameter when running
clustSIGNAL will generate different entropy values. Comparing entropy values
from different neighbourhood sizes could be useful in estimating the size of the
domain-like regions.

## Summary

::: summary
-   ClustSIGNAL is a method developed to counter data sparsity by performing an
    adaptive smoothing approach to generate new gene expression that is used for
    clustering.

-   The method comes with an ensemble of options that allow selection of
    neighbourhood, weighting, batch correction, and clustering parameters.

-   ClustSIGNAL can be run on datasets with mulitple samples and will identify
    clusters in all samples in one run.

-   ClustSIGNAL uses both gene expression and spatial information to generate
    clusters from subcellular spatial transcriptomics data by stabilising the
    expression of cells in homogeneous regions and preventing the loss of
    distinct expression of cells in more heterogeneous regions.

-   Neighbourhood entropy values generated by clustSIGNAL can be used for data
    exploration to estimate the extent of "domainness" in tissue samples.
:::

<details>

<summary>**Session Information**</summary>

```{r}
sessionInfo()
```

</details>
